### ISSUE TITLE
segment stuck and always failing

### ISSUE THREAD

**Issue by vrischmann** [thelastpickle/cassandra-reaper#376]
Hi,

I noticed today there's a segment stuck since approximately 5 days, the fail count is at 661 now.

The only exceptions I see in the log are these:

    DEBUG  [2018-03-22 16:39:26,393] [pool-9-thread-7] i.c.j.JmxProxy - No repair has run yet on the node. Ignoring exception.
    javax.management.InstanceNotFoundException: org.apache.cassandra.internal:type=AntiEntropySessions
            at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.getMBean(DefaultMBeanServerInterceptor.java:1095)
            at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.getAttribute(DefaultMBeanServerInterceptor.java:643)
            at com.sun.jmx.mbeanserver.JmxMBeanServer.getAttribute(JmxMBeanServer.java:678)
            at javax.management.remote.rmi.RMIConnectionImpl.doOperation(RMIConnectionImpl.java:1445)
            at javax.management.remote.rmi.RMIConnectionImpl.access$300(RMIConnectionImpl.java:76)
            at javax.management.remote.rmi.RMIConnectionImpl$PrivilegedOperation.run(RMIConnectionImpl.java:1309)
            at javax.management.remote.rmi.RMIConnectionImpl.doPrivilegedOperation(RMIConnectionImpl.java:1401)
            at javax.management.remote.rmi.RMIConnectionImpl.getAttribute(RMIConnectionImpl.java:639)
            at sun.reflect.GeneratedMethodAccessor351.invoke(Unknown Source)
            at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	    ...

there are quite a lot of them but from what I remember these exceptions should be harmless so I'm not sure if it's relevant.

Otherwise I noticed a couple of things which seem weird to me. Here it looks like the segment has finished but yet is detected has having timed out ?

    DEBUG  [2018-03-22 15:00:54,424] [bep_cluster] i.c.j.JmxProxy - Received notification: javax.management.Notification[source=repair:10385][type=progress][message=Repair command #10385 finished in 3 minutes 25 seconds] with type progress
    DEBUG  [2018-03-22 15:00:54,424] [bep_cluster] i.c.j.JmxProxy - Handling notification: javax.management.Notification[source=repair:10385][type=progress][message=Repair command #10385 finished in 3 minutes 25 seconds] with repair handler true
    DEBUG  [2018-03-22 15:00:54,425] [bep_cluster:00000000-0000-0250-0000-000000000000:00000000-003a-b11c-0000-000000000000] i.c.s.SegmentRunner - handle called for repairCommandId 10385, outcome Optional.absent() / Optional.of(COMPLETE) and message: Repair command #10385 finished in 3 minutes 25 seconds
    DEBUG  [2018-03-22 15:00:54,426] [bep_cluster:00000000-0000-0250-0000-000000000000:00000000-003a-b11c-0000-000000000000] i.c.s.SegmentRunner - repair session finished for segment with id '00000000-003a-b11c-0000-000000000000' and repair number '10385'
    INFO   [2018-03-22 15:00:54,427] [bep_cluster:00000000-0000-0250-0000-000000000000:00000000-003a-b11c-0000-000000000000] i.c.s.SegmentRunner - Repair command 10385 on segment 00000000-003a-b11c-0000-000000000000 returned with state NOT_STARTED
    INFO   [2018-03-22 15:00:54,427] [bep_cluster:00000000-0000-0250-0000-000000000000:00000000-003a-b11c-0000-000000000000] i.c.s.SegmentRunner - Repair command 10385 on segment 00000000-003a-b11c-0000-000000000000 never managed to start within timeout.
    INFO   [2018-03-22 15:00:54,428] [bep_cluster:00000000-0000-0250-0000-000000000000:00000000-003a-b11c-0000-000000000000] i.c.s.SegmentRunner - Postponing segment 00000000-003a-b11c-0000-000000000000

Here's a second one where the segment appears to be cancelled just after it started:

    INFO   [2018-03-22 16:37:56,264] [bep_cluster:00000000-0000-0250-0000-000000000000:00000000-003a-b11c-0000-000000000000] i.c.s.SegmentRunner - Repair for segment 00000000-003a-b11c-0000-000000000000 started, status wait will timeout in 60000 millis
    DEBUG  [2018-03-22 16:37:56,266] [bep_cluster:00000000-0000-0250-0000-000000000000:00000000-003a-b11c-0000-000000000000] i.c.s.SegmentRunner - updated segment 00000000-003a-b11c-0000-000000000000 with state RUNNING
    INFO   [2018-03-22 16:38:56,265] [bep_cluster:00000000-0000-0250-0000-000000000000:00000000-003a-b11c-0000-000000000000] i.c.s.SegmentRunner - Repair command 8017 on segment 00000000-003a-b11c-0000-000000000000 returned with state RUNNING
    INFO   [2018-03-22 16:38:56,265] [bep_cluster:00000000-0000-0250-0000-000000000000:00000000-003a-b11c-0000-000000000000] i.c.s.SegmentRunner - Repair command 8017 on segment 00000000-003a-b11c-0000-000000000000 has been cancelled while running
    INFO   [2018-03-22 16:38:56,266] [bep_cluster:00000000-0000-0250-0000-000000000000:00000000-003a-b11c-0000-000000000000] i.c.s.SegmentRunner - Postponing segment 00000000-003a-b11c-0000-000000000000
    INFO   [2018-03-22 16:38:56,308] [bep_cluster:00000000-0000-0250-0000-000000000000:00000000-003a-b11c-0000-000000000000] i.c.s.SegmentRunner - Aborting repair on segment with id 00000000-003a-b11c-0000-000000000000 on coordinator 10.51.0.8
    DEBUG  [2018-03-22 16:38:56,311] [bep_cluster:00000000-0000-0250-0000-000000000000:00000000-003a-b11c-0000-000000000000] i.c.s.SegmentRunner - Exiting synchronized section with segment ID 00000000-003a-b11c-0000-000000000000
    DEBUG  [2018-03-22 16:38:56,312] [bep_cluster:00000000-0000-0250-0000-000000000000:00000000-003a-b11c-0000-000000000000] i.c.s.RepairRunner - In repair run #00000000-0000-0250-0000-000000000000, triggerRepair on segment 00000000-003a-b11c-0000-000000000000 ended with state NOT_STARTED
    DEBUG  [2018-03-22 16:39:00,967] [bep_cluster:00000000-0000-0250-0000-000000000000:00000000-003a-b11c-0000-000000000000] i.c.s.SegmentRunner - Run repair for segment #00000000-003b-4469-0000-000000000000

I don't know much about the internals of reaper so I may be misinterpreting the logs.

Finally here's a screenshot from the web ui showing the lone segment that always fail.

![screenshot from 2018-03-22 17-49-49](https://user-images.githubusercontent.com/1916079/37785092-8fb6518e-2df9-11e8-8610-eb5ebb5ebb21.png)

You can see the other segments have finished on the 17th.

I tried to:
 * abort/restart the segment
 * restart reaper

but it didn't change anything.

**Comment by vrischmann** [comment]
To be honest I don't remember exactly why it's set that low, it's a configuration that was made a couple years ago already.  The settings haven't changed in between, no.

Since I opened the issue I haven't seen this issue again but at some point I drop the data I already had in postgresql and switched to the cassandra backend starting from scratch.

Since then I haven't had any problems, but I increased the timeout anyway just in case.

If I ever encounter this issue again i'll come back.
